File: yntra-ios/YntraIOS/ViewModels/DirectoryViewModel.swift
import SwiftUI
import yntra_core

@MainActor
class DirectoryViewModel: ObservableObject {
    @Published var users: [WorkspaceUser] = []
    @Published var teams: [Team] = []
    @Published var errorMessage: String? = nil
    
    private var observer: SwiftDbObserver?

    init() {
        refreshDirectory()
        
        self.observer = SwiftDbObserver { [weak self] in
            self?.refreshDirectory()
        }
        registerObserver(observer: self.observer!)
    }

    func refreshDirectory() {
        Task {
            do {
                self.users = try await getUsers(requesterUserId: "user-1")
                self.teams = try await getTeams(requesterUserId: "user-1")
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func inviteUser(email: String, name: String, role: String) {
        Task {
            do {
                _ = try await inviteUserViaDirectory(
                    requesterUserId: "user-1",
                    workspaceId: "workspace-1",
                    email: email,
                    name: name,
                    role: role
                )
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func createTeam(name: String) {
        Task {
            do {
                _ = try await addTeamViaDirectory(
                    requesterUserId: "user-1",
                    workspaceId: "workspace-1",
                    name: name
                )
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func deleteWorkspaceUser(userId: String) {
        Task {
            do {
                try await deleteUser(requesterUserId: "user-1", userId: userId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
}

File: yntra-android/app/src/main/java/com/yntra/app/viewmodels/DirectoryViewModel.kt
package com.yntra.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import uniffi.yntra_core.*
import com.yntra.app.KotlinDbObserver

class DirectoryViewModel : ViewModel() {
    private val _users = MutableStateFlow<List<WorkspaceUser>>(emptyList())
    val users: StateFlow<List<WorkspaceUser>> = _users

    private val _teams = MutableStateFlow<List<Team>>(emptyList())
    val teams: StateFlow<List<Team>> = _teams

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private val observer = KotlinDbObserver {
        refreshDirectory()
    }

    init {
        refreshDirectory()
        registerObserver(observer)
    }

    fun refreshDirectory() {
        viewModelScope.launch {
            try {
                _users.value = getUsers(com.yntra.app.SessionManager.activeUserId)
                _teams.value = getTeams(com.yntra.app.SessionManager.activeUserId)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun inviteUser(email: String, name: String, role: String) {
        viewModelScope.launch {
            try {
                inviteUserViaDirectory(
                    requesterUserId = com.yntra.app.SessionManager.activeUserId,
                    workspaceId = com.yntra.app.SessionManager.activeWorkspaceId,
                    email = email,
                    name = name,
                    role = role
                )
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun createTeam(name: String) {
        viewModelScope.launch {
            try {
                addTeamViaDirectory(
                    requesterUserId = com.yntra.app.SessionManager.activeUserId,
                    workspaceId = com.yntra.app.SessionManager.activeWorkspaceId,
                    name = name
                )
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun deleteWorkspaceUser(userId: String) {
        viewModelScope.launch {
            try {
                deleteUser(com.yntra.app.SessionManager.activeUserId, userId)
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        clearObservers()
    }
}
